### ISSUE TITLE
glm5: a FULL-COVER prefix-cache hit routes the request to the plain decode path (route=plain, no DFlash2), halving speed for repeated prompts

### ISSUE THREAD

**Issue by avifenesh** [avifenesh/memra#74]
Observed on the glm53 ship box (Texas, vast 49585342, memra-0.123.0-6a23ef199612, PP3 on GPUs 1-3, MEMRA_PREFIX_LATENT + MEMRA_HYPER_SUFFIX_PRIME + MEMRA_GLM5_SPEC_PREFIX, MEMRA_GLM5_SPEC=1, PMIN 0.7, MEMRA_PREFIX_CACHE_MB=8192), 2026-09-02 06:59-07:01Z, single stream, no other traffic (ledger checked).

Six identical vendor-default requests (67-token digits prompt, max_tokens 512, streamed). Rep 0 (cold) took 8.4 s = 58 tok/s wall, [glm5-spec] route=spec. Reps 1-5, each a full-cover cache hit, took 16.6-16.8 s = 28-30 tok/s wall. Slot log for every one of them:

```
[worker] plain-affinity: declined (no checkpoint retained; 2 parked, 67 prompt tokens; model zai/glm-5.3-flash)
[prefix-cache] hit: 67 of 67 prompt tokens from cache (model zai/glm-5.3-flash)
[glm5-spec] route=plain K=3 model="zai/glm-5.3-flash" tenant="t:..." prompt=67 wave=1 sampled=1 penalized=0 cold=0 restored=0
```

So when the prompt is served entirely from the prefix cache (`hit: N of N`), the spec router answers `route=plain` with `cold=0 restored=0`: the drafter is not engaged and decode runs at the plain speed (~30 tok/s on this shape vs ~66-70 tok/s decode-only with DFlash2). Strict-prefix hits with a suffix (multi-turn) were not isolated in this cell; the soak's cache_mt class measured content, not speed, on those.

Why no instrument caught it before exposure: the claim cell (claim_cell.sh) and the reverify wave (reverify_wave.py) both send a fresh `cache_salt` per request to measure the cold-prefix ceiling, and the soak's spec probes used unique prompts. A repeated byte-identical prompt is exactly what retries, eval harnesses and provider latency probes send.

Receipts: darklanes research/glm5-serving-launch-20260901/soak-20260901/receipts/go-live/20260902T065928Z/publicity-cell-rig-x5.json (identical prompts: wall median 29.46, decode median 30.8) vs publicity-cell-rig-x5-nonce.json (fresh nonce per prompt: wall median 57.75, decode median 69.7), same minute, same vantage, same fingerprint.

Related: #34 (partial-hit continuation divergence and prompt-shape-dependent engagement). Ask: on a full-cover hit, either take the spec path from the restored state (the drafter needs the tail context it already has) or state in the route line WHY plain was chosen so operators can size the cost. Until then the cache-read row is honest (cached_tokens bills correctly) but a cache hit costs the customer half the decode speed on this shape.

**Comment by avifenesh** [comment]
CLAIMED glm (coordinator session) 2026-09-02T14:24:11Z branch lane/glm5-fullcover-spec-route-20260902

**Comment by avifenesh** [comment]
Landed in #88 (merged 3a3064f3d).

**Mechanism, attributed.** Two gates from the parent lane's restore half. (1) `worker.rs:18521` put the ENTIRE glm5 carrier probe behind `suffix_len >= PRIME_MIN_T`, so a full-cover hit (`suffix_len == 0`) never reached it: `glm5_prefix_restored_dkv` stayed None, `glm5_restored_carrier` was false, and `glm5_cold` was ALSO false because the hit put a cache on the session, so `glm5_route_admits`' `cold_or_restored_carrier` term failed and the request routed plain with no reason on the line. (2) `glm_spec.rs:2052` refused an empty suffix by construction and `worker.rs:22131` finished a session whose prefill queue drained empty. Not (a): the restored state carries everything the drafter needs, and the MTP family has served this exact shape since lane/spec-cache (`spec_restore_refusal` admits full cover on boundary hidden + logits). The refusal's stated premise ("the plain boundary-logits resume is faster than any prime") is true of PREFILL and silent about DECODE, which is where the whole cost was.

**Shipped now, default-on:** the `[glm5-spec]` route line carries `reason=<token>` on every `route=plain`, from a predicate pinned equivalent to `glm5_route_admits` over the whole class matrix. The live line would read `route=plain ... cold=0 restored=0 reason=full-cover-hit`. That closes the "operators can size the cost" half unconditionally.

**Shipped behind `MEMRA_GLM5_SPEC_FULLCOVER`, DEFAULT OFF:** the full-cover hit becomes a spec carrier (no prime, no tap rows, drafter ctx KV from the entry tail, anchor off the entry's boundary row). Default OFF is the new-flags law, not caution: the lane had no GPU, so the arm has no receipt.

**Still needs a GPU before the flag flips anywhere.** Three cells, pre-stated in the FLAGS.md row:
1. rig, exactness only: `cargo test -p memra-engine --test glm5_dflash_session_gpu -- --ignored gpu_full_cover_restore_bytes_match_plain_decode_and_cold_acceptance` under `flock /tmp/memra-5090.lock` — greedy tape byte-identical to plain decode, acceptance identical to a cold session, and the sampled twin equal to a cold sampled session at one seed;
2. serving box: this issue's own six-identical-prompts probe re-run with the flag armed, interleaved x5 against the flag-off arm, expecting `route=spec ... restored=1`, `[glm5-spec] RESTORED session: ... arm full-cover`, a `[glm5-acc]` line, and decode recovering toward the cold ceiling;
3. the owner-law 8-turn larger-prompt cache twin.

Leaving this open for the deploy lane to close on those receipts, per the issues-close-on-serving law: the engine work is merged but nothing is serving differently yet.

https://claude.ai/code/session_01UxWJVX4keng1f17h4fGUyi

**Comment by avifenesh** [comment]
The three flip-condition cells are run. Rented box, owner-authorised: **vast 49659267**, 4x RTX PRO
6000 Blackwell WS 600 W, Alberta, destroyed at close. No production box was touched: model bytes out
of R2 only (34/34 sha256-verified against the darklanes artifact registry), every request to
`127.0.0.1` on the rented box, the glm5 ship launcher unmodified.

Engine `3a3064f3db77` (main with #88) + darklanes `2b0e6d928`, binary sha256
`33b47437f778087b964cc099474f23a1aeda9bb573174a99669e6bc4b30a1203`.

## Result: two of three green, and the flag still stays OFF

**1. Exactness: GREEN, twice.** `gpu_full_cover_restore_bytes_match_plain_decode_and_cold_acceptance`
(gate 13) passes on this hardware:

```
gate 13 PASS: full-cover restore == plain bytes over 20 tokens, acceptance 0 / 57 identical to
cold; sampled twin == cold sampled over 20 tokens at seed 0x5eed
```

And at the serving waist on the real 3-card PP3 shape, one greedy tape sha `eef4c48ada022999` covers
the cold request and both full-cover hits in **both** arms:

| arm | req | route line | cached | wall s | tape |
|---|---|---|---|---|---|
| OFF | cold | `route=spec ... cold=1 restored=0 reason=-` | 0/67 | 3.598 | `eef4c48a…` |
| OFF | hit | `route=plain ... reason=full-cover-hit` | 67/67 | 7.009 | `eef4c48a…` |
| OFF | hit | `route=plain ... reason=full-cover-hit` | 67/67 | 7.028 | `eef4c48a…` |
| ON | cold | `route=spec ... cold=1 restored=0 reason=-` | 0/67 | 4.993 | `eef4c48a…` |
| ON | hit | `route=spec ... cold=0 restored=1 reason=-` | 67/67 | 3.404 | `eef4c48a…` |
| ON | hit | `route=spec ... restored=1` | 67/67 | 3.405 | `eef4c48a…` |

Stated rather than rounded away: the restored session's first burst accepts **21/22** where the cold
session accepts **22/22**, and the two `[glm5-acc]` series then run one token apart for the whole
generation, under a byte-identical tape. Gate 13's acceptance-identity assertion passes on its
fixture; on the real checkpoint the restored session drops exactly one draft in the first burst.

**3. The owner-law 8-turn twin: GREEN, and the flag does no harm.** 80 turns across both arms, zero
non-completing, `cached_tokens > 0` on every warm turn, TTFT flat at 0.95-1.06 s while the prompt
grows from 2.4k to 3.7k, no snapshot refusals, no panic. Worth being explicit about what this cell
can and cannot show: every warm turn appends a new user message, so turns 2-8 are strict-prefix hits
with a non-empty suffix, i.e. the parent lane's suffix-prime restore, which this flag does not touch.
It is the no-harm check, and it also clears the live prod restore path of the failure below.

**2. Your own repeated-prompt cell: FAILED, on completion rather than on speed.**

With the flag armed, a full-cover RESTORED session panics the GPU worker:

```
thread 'memra-gpu-worker' panicked at crates/memra-engine/src/dflash.rs:804:13:   (sampled walk)
thread 'memra-gpu-worker' panicked at crates/memra-engine/src/dflash.rs:221:13:   (greedy walk)
walk: candidate 4294967295 outside codebook vocab
[worker] FATAL: worker unrecoverable after 1 respawn attempt(s) - exiting the process
```

`4294967295` is `u32::MAX` read out of the drafter's top-k candidate buffer. On the **vendor-default
sampled** shape, which is the shape this issue is about, **18 of 25 full-cover hits did not
complete**. The flag-off arm lost **0 of 50**. One respawn and then the process exits, so this is
fleet-fatal, not a degraded request. Filed with its full characterisation as #95.

## The prize, where it completes, is exactly what #88 predicted

Six identical vendor-default requests, 67-token digits prompt, 512 max_tokens, streamed, no
`cache_salt`, interleaved x5 with one fresh boot per cell per arm and the arm proved per boot off
`/proc/<pid>/environ`:

| shape | full-cover hit decode OFF -> ON | ratio | hit as a fraction of the arm's OWN cold ceiling |
|---|---|---|---|
| vendor-default sampled | 35.945 -> 83.354 tok/s | **2.319x** | OFF **0.430x** -> ON **1.004x** |
| greedy (instrument only) | 36.642 -> 87.160 tok/s | **2.379x** | OFF **0.417x** -> ON **0.993x** |

That last column is the cleanest statement of the defect and of the fix: with the flag off a
full-cover cache hit decodes at **43% of what the same box's cold request decodes**; with it on the
hit decodes at the cold ceiling.

**These absolute numbers are not publishable.** This is not the prod host class (EPYC 9B14,
Workstation cards, 20M-iteration single-thread loop 1.035-1.049 s vs prod's Xeon 6767P), and a
host-thread-bound box exaggerates a speculative-decoding win because the plain path pays more host
time per token. The green ratio is an **upper bound** on prod, not a prediction of it.

## Two things that are settled

- **The defect in this issue is independently reproduced** on a second box and a second memra SHA:
  the flag-off arm answers `route=plain ... cold=0 restored=0 reason=full-cover-hit` on 50 of 50
  full-cover hits, `cached_tokens = prompt_tokens = 67` on every one, at 0.43x of its own cold
  decode. That half needs no further measurement.
- **The unconditional half of #88 works and is worth keeping as shipped.** `reason=full-cover-hit`
  on the route line is exactly the line an operator did not have when this shipped, and it cost
  nothing.

## Ask

Keep `MEMRA_GLM5_SPEC_FULLCOVER` OFF everywhere until #95 lands, then re-run cell 2 only (cells 1 and
3 are banked green and do not need a second box). The FLAGS.md row's "Receipts: NONE YET, by
construction" sentence should now be replaced rather than deleted: conditions 1 and 3 are green with
named receipts, condition 2 is red with a named cause.

Full receipts, per-request JSON, per-boot arm identity, three banked server logs and every harness
script: darklanes `research/glm5-fullcover-measure-20260902/`, PR avifenesh/darklanes#132.

https://claude.ai/code/session_01UxWJVX4keng1f17h4fGUyi
